tests/regression/src/lib.rs

```diff
@@ -55,24 +79,228 @@ mod tests {
             cg::cachegrind::start_instrumentation();
         }
     }
-    /// Environment variable to determine whether to run under valgrind or solely test functionality.
-    fn is_running_under_valgrind() -> bool {
-        env::var("ENABLE_VALGRIND").is_ok()
+
+    #[derive(Debug)]
+    enum RegressionTestMode {
+        Default,
+        Valgrind,
+        Diff,
+    }
+
+    impl RegressionTestMode {
+        fn from_env() -> Self {
+            if env::var("ENABLE_VALGRIND").is_ok() {
+                RegressionTestMode::Valgrind
+            } else if env::var("DIFF_MODE").is_ok() {
+                RegressionTestMode::Diff
+            } else {
+                RegressionTestMode::Default
+            }
+        }
     }
 
     fn valgrind_test<F>(test_name: &str, test_body: F) -> Result<(), s2n_tls::error::Error>
     where
         F: FnOnce(&InstrumentationControl) -> Result<(), s2n_tls::error::Error>,
     {
-        if !is_running_under_valgrind() {
-            let ctrl = InstrumentationControl;
-            test_body(&ctrl)
-        } else {
-            run_valgrind_test(test_name);
-            Ok(())
+        match RegressionTestMode::from_env() {
+            RegressionTestMode::Valgrind => {
+                let raw_profile = RawProfile::new(test_name);
+                AnnotatedProfile::new(&raw_profile);
+            }
+            RegressionTestMode::Diff => {
+                let (prev_profile, curr_profile) = RawProfile::query(test_name);
+                DiffProfile::new(&prev_profile, &curr_profile).assert_performance();
+            }
+            RegressionTestMode::Default => {
+                let ctrl = InstrumentationControl;
+                test_body(&ctrl)?
+            }
+        };
+        Ok(())
+    }
+
+    struct RawProfile {
+        test_name: String,
+        commit_hash: String,
+    }
+
+    impl RawProfile {
+        fn new(test_name: &str) -> Self {
+            let commit_hash = git::get_current_commit_hash();
+            create_dir_all(format!("target/{commit_hash}")).unwrap();
+
+            let raw_profile = Self {
+                test_name: test_name.to_owned(),
+                commit_hash,
+            };
+
+            let mut command = Command::new("valgrind");
+            command
+                .args([
+                    // use cachegrind to get instruction count
+                    "--tool=cachegrind",
+                    // write output to output file
+                    &format!("--cachegrind-out-file={}", raw_profile.path()),
```

Comment:
can these be split into different arguments?
```suggestion
                    "--cachegrind-out-file",
                    raw_profile.path(),
```

Reply:
No, that wouldn't work. The --cachegrind-out-file={} command is an option provided by cachegrind and requires that you specify the output file path for it to dump contents too in that format as one argument. Otherwise, it generates an output file based on the Process ID.